File: packages/sidecar/src/crash_reporter.ts
import * as Sentry from "@sentry/node";

export interface CrashReporterConfig {
  dsn?: string;
  environment?: string;
  release?: string;
  userOptIn?: boolean;
}

let initialized = false;

export function initCrashReporter(config: CrashReporterConfig): void {
  if (initialized) return;
  if (!config.dsn) return;

  Sentry.init({
    dsn: config.dsn,
    environment: config.environment ?? process.env["NODE_ENV"] ?? "production",
    release: config.release ?? "0.1.0",
    tracesSampleRate: 0.2,
    profilesSampleRate: 0.1,
    sendDefaultPii: false,
  });

  initialized = true;
}

export function captureException(
  error: unknown,
  context?: Record<string, unknown>
): void {
  if (!initialized) return;

  Sentry.captureException(error, (scope) => {
    if (context) {
      for (const [key, value] of Object.entries(context)) {
        scope.setTag(key, String(value));
      }
    }
    return scope;
  });
}

export function captureMessage(
  message: string,
  level?: "info" | "warning" | "error" | "fatal"
): void {
  if (!initialized) return;

  const sentryLevel = level
    ? {
        info: "info" as const,
        warning: "warning" as const,
        error: "error" as const,
        fatal: "fatal" as const,
      }[level] ?? "info"
    : "info";

  Sentry.captureMessage(message, sentryLevel);
}

export function setUserContext(
  userId?: string,
  email?: string,
  username?: string
): void {
  if (!initialized) return;

  Sentry.setUser({
    id: userId,
    email: email,
    username: username,
  });
}

export function clearUserContext(): void {
  if (!initialized) return;
  Sentry.setUser(null);
}

export function addBreadcrumb(
  message: string,
  category?: string,
  level?: "debug" | "info" | "warning" | "error"
): void {
  if (!initialized) return;

  Sentry.addBreadcrumb({
    message,
    category: category ?? "app",
    level: level ?? "info",
  });
}

export async function close(): Promise<void> {
  if (!initialized) return;
  await Sentry.close();
}

export async function readDsnFromFile(path: string): Promise<string> {
  try {
    const fs = await import("node:fs/promises");
    const content = await fs.readFile(path, "utf-8");
    return content.trim();
  } catch {
    return "";
  }
}

File: apps/liminal_desktop/lib/core/crash_reporter.dart
import 'dart:io';

import 'package:sentry_flutter/sentry_flutter.dart';

class CrashReporter {
  static CrashReporter? _instance;
  static CrashReporter get instance => _instance ??= CrashReporter._();

  CrashReporter._();

  bool _initialized = false;
  bool _userOptIn = false;

  Future<void> init({
    required String dsn,
    String environment = 'production',
    String release = '0.1.0',
    bool userOptIn = false,
  }) async {
    if (_initialized) return;

    _userOptIn = userOptIn;

    await SentryFlutter.init(
      (options) {
        options.dsn = dsn;
        options.environment = environment;
        options.release = release;
        options.tracesSampleRate = 0.2;
        options.profilesSampleRate = 0.1;
        options.attachStacktrace = true;
        options.reportSilentFlutterErrors = true;
        options.enableUserInteractionBreadcrumbs = false;
        options.sendDefaultPii = false;
      },
    );

    _initialized = true;
    _captureNativeCrashes();
  }

  void _captureNativeCrashes() {
    if (_userOptIn) {
      Sentry.captureMessage('Crash reporter initialized');
    }
  }

  Future<void> setUserContext({
    String? userId,
    String? email,
    String? username,
  }) async {
    if (!_initialized || !_userOptIn) return;

    final user = SentryUser(
      id: userId ?? _generateAnonymousId(),
      email: email,
      username: username,
    );

    Sentry.configureScope((scope) {
      scope.setUser(user);
    });
  }

  Future<void> clearUserContext() async {
    Sentry.configureScope((scope) {
      scope.setUser(null);
    });
  }

  void addBreadcrumb({
    required String message,
    String? category,
    String? level,
  }) {
    Sentry.addBreadcrumb(
      Breadcrumb(
        message: message,
        category: category ?? 'app',
        level: _parseLevel(level),
      ),
    );
  }

  SentryLevel _parseLevel(String? level) {
    return switch (level) {
      'debug' => SentryLevel.debug,
      'info' => SentryLevel.info,
      'warning' => SentryLevel.warning,
      'error' => SentryLevel.error,
      'fatal' => SentryLevel.fatal,
      _ => SentryLevel.info,
    };
  }

  Future<void> captureException(
    dynamic exception,
    StackTrace? stackTrace, {
    String? context,
  }) async {
    if (!_initialized) return;

    await Sentry.captureException(
      exception,
      stackTrace: stackTrace,
      withScope: (scope) {
        if (context != null) {
          scope.setTag('context', context);
        }
      },
    );
  }

  Future<void> captureMessage(String message, {String? level}) async {
    if (!_initialized) return;

    await Sentry.captureMessage(
      message,
      level: _parseLevel(level),
    );
  }

  String _generateAnonymousId() {
    return 'anon-${DateTime.now().millisecondsSinceEpoch}';
  }

  static bool get isOptedIn => instance._userOptIn;

  static Future<String> readDsnFromFile(String path) async {
    final file = File(path);
    if (!await file.exists()) {
      return '';
    }
    final content = await file.readAsString();
    return content.trim();
  }
}
